shcomplete/model2correct.py

```diff
@@ -0,0 +1,224 @@
+        if reduction:
+            if inverted:
+                begin_of_command = np.amin(np.nonzero(X))
+                for i in range(begin_of_command, X.shape[0]):
+                    command += self.indices_char[X[i]]
+            else:
+                X = np.squeeze(X)
+                end_of_command = np.amax(np.nonzero(X))
+                for i in range(end_of_command + 1):
+                    command += self.indices_char[X[i]]
+            return command
+        else:
+            try:
+                end_of_command = np.amax(np.nonzero(X))
+            except ValueError:
+                end_of_command = -1
+            for i in range(end_of_command + 1):
+                try:
+                    command += self.indices_char[X[i]]
+                except KeyError:
+                    pass
+        return command
+
+
+def generator(chars, vocabulary, corpus, max_cmd_len, batch_size, level_noise):
+    """
+    Generate a tuple (misspelled_command, true_command) to train the model.
+    The misspelled_command is generated by adding artificial mistakes.abs
+    Padding is added to feed the model with fixed length sequences of characters.
+    An epoch finishes when samples_per_epoch samples have been seen by the model.
+    """
+    seq2seq = Seq2seq(chars)
+    vocab = Vocabulary(vocabulary)
+    vocab_trie = vocab.trie(vocabulary)
+    with open(corpus) as f:
+        lines = np.array([cmd.strip() for cmd in f if cmd != "\n"])
+        nb_lines = lines.shape[0]
+        id_lines = np.arange(nb_lines)
+        while True:
+            commands = []
+            misspelled_commands = []
+            while len(commands) < batch_size:
+                command = lines[np.searchsorted(id_lines, randint(nb_lines - 1))]
+                prefix = vocab_trie.longest_prefix(command)[0]
+                if prefix is not None:
+                    misspelled_command = create_mistakes(chars, prefix, level_noise)
+                    commands.append(prefix)
+                    misspelled_commands.append(misspelled_command[::-1])
+            misspelled_commands_encoded = seq2seq.encode(misspelled_commands)
+            commands_encoded = seq2seq.encode(commands)
+            X = pad_sequences(misspelled_commands_encoded, maxlen=max_cmd_len,
+                              dtype='int32', padding='pre', truncating='pre')
+            y = pad_sequences(commands_encoded, maxlen=max_cmd_len,
+                              dtype='int32', padding='post', truncating='post')
+            yield X, y[:, :, np.newaxis]
+
+
+def generate_model(chars, args):
+    """
+    Generate the model.
+    """
+    nb_chars = len(chars) + 1
+    emb_weights = np.eye(nb_chars)
+
+    model = Sequential()
+    model.add(Embedding(input_dim=nb_chars, output_dim=nb_chars, input_length=args.max_cmd_len,
+                        weights=[emb_weights], trainable=False))
+    for layer_id in range(args.input_layers):
+        model.add(recurrent.LSTM(args.hidden_layers,
```

Comment:
Please add the ability to have GRU instead of LSTM. Actually, always try to experiment with GRUs from the very beginning and use LSTMs only as a reference in the end. The reason is speed.

This can be added *after* this PR is merged.